django/db/backends/ddl_references.py

```diff
@@ -198,3 +201,56 @@ def rename_column_references(self, table, old_column, new_column):
 
     def __str__(self):
         return self.template % self.parts
+
+
+class Expressions(Reference):
+    def __init__(self, table, expressions, compiler, quote_value, opclasses):
+        self.table = table
+        self.compiler = compiler
+        self.expressions = [
+            expression.resolve_expression(self.compiler.query) for expression in expressions
+        ]
+        self.quote_value = quote_value
+        self.opclasses = opclasses
+
+    def references_table(self, table):
+        return self.table == table
+
+    def references_column(self, table, column):
+        columns = (col.target.column for col in self.compiler.query._gen_cols(self.expressions))
+        return self.table == table and column in columns
+
+    def rename_table_references(self, old_table, new_table):
+        if self.table == old_table:
+            self.table = new_table
+
+    def rename_column_references(self, table, old_column, new_column):
+        if self.table != table:
+            return
+        expressions = [deepcopy(expression) for expression in self.expressions]
+        for col in self.compiler.query._gen_cols(expressions):
+            if col.target.column == old_column:
+                col.target.column = new_column
+        self.expressions = expressions
+
+    def __str__(self):
+        sql_expressions = []
+        for index, expression in enumerate(self.expressions):
+            ordering = ''
+            if expression.ordered:
+                ordering = ' DESC' if expression.descending else ' ASC'
+                expression = expression.get_source_expressions()[0]
+            sql, params = self.compiler.compile(expression)
+            try:
+                opclass = ' ' + self.opclasses[index]
+            except IndexError:
+                opclass = ''
+            if isinstance(expression, Col):
+                # MySQL does not support functional indexes on columns
+                template = '%s%s%s'
```

Comment:
This seems out of place. Is this branch really specific to MySQL? Is there a way we could avoid the `Col` import in the first place.

Reply:
MySQL will throw the error “Functional index on a column is not supported. Consider using a regular index instead.”  when you try to create an index on a column wrapped with parentheses, i.e `CREATE INDEX foo ON bar ((name), (UPPER(name));`. 

I guess that it’s not really specific to MySQL as the wrapping is unnecessary on other engines and having columns wrapped with parentheses is quite unexpected for users.  We should probably remove that comment. I would love to get rid of the `isinstance`-check. I’m not aware of any trick which might tell if an `Expression` is only a column reference, I would have to take a deeper dive into the expressions-code.

Reply:
> I guess that it’s not really specific to MySQL as the wrapping is unnecessary on other engines and having columns wrapped with parentheses is quite unexpected for users. We should probably remove that comment. I would love to get rid of the isinstance-check. I’m not aware of any trick which might tell if an Expression is only a column reference, I would have to take a deeper dive into the expressions-code.

Could we do that at the `Index.create_sql` level by wrapping `F`'s into `Func(f_expr, template='(%(expressions)s)')`?

Reply:
Isn’t that what we’re trying to avoid? Only expressions should be wrapped by parentheses. But we should be able to do it  for non-`F`’s instead! That should get rid of the `if`-statement which chooses a template depending on `Col` in ddl `Expressions`

Reply:
Looks like it does not work for ordered expressions, i.e `Index(Lower("name").desc(), "email",  ..)` would generate the following SQL by using `Func(expr, template='(%(expressions)s)')`: 
`CREATE INDEX foo ON bar ((LOWER("name") DESC), "email")` which is invalid SQL, the ordering (and opclass) needs to be kept outside the wrapping parentheses: `CREATE INDEX foo ON bar ((LOWER("name")) ASC, "email")`

Reply:
Maybe this is something that we can solve when adding support for `OpClass`?

Reply:
I think that most of the expression special casing and resolving should be done at the `Index.create_sql` level. The only purpose of `ddl_references` is to hold references to identifiers and allow renaming if necessary, it shouldn't have any knowledge about `django.db.models` abstractions.